### PR TITLE
[RISC-V] Test TotalOrderIeee754ComparerTests: Fix NFloat - NaN testcases for RISC-V

### PR DESCRIPTION
**Pull request by denis-paranichev**
float->double cast instruction produced by NFloat constructor does not preserve NaN sign and payload on RISC-V.
That's why all NaN values in test data are same positive NaNs without payload.

Part of https://github.com/dotnet/runtime/issues/84834
cc @tannergooding @gbalykov @t-mustafin @clamp03 @tomeksowi

### CODE REVIEW

**Comment by denis-paranichev** [comment]
@dotnet-policy-service agree company="Samsung"

**Review by gbalykov — APPROVED** [review]

**Review by tomeksowi — APPROVED** [review]

**Review by clamp03 — APPROVED** [review]

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Numerics/TotalOrderIeee754ComparerTests.cs [hunk 1]
```diff
@@ -94,8 +94,38 @@ public void TotalOrderTestHalf(Half x, Half y, int result)
...
+                yield return new object[] { BitConverter.UInt32BitsToSingle(0x7FC00000), 1.0f, 1 };
+                yield return new object[] { BitConverter.UInt32BitsToSingle(0x7FC00000), float.PositiveInfinity, 1 };
+                yield return new object[] { float.NaN, float.NaN, 0 };
+                if (PlatformDetection.IsRiscV64Process) // float->double cast does not preserve NaN payload and sign on RISC-V
+                {
+                    yield return new object[] { BitConverter.UInt32BitsToSingle(0xFFC00000), float.NegativeInfinity, 1 };
+                    yield return new object[] { BitConverter.UInt32BitsToSingle(0xFFC00000), -1.0f, 1 };
+                    yield return new object[] { BitConverter.UInt32BitsToSingle(0xFFC00000), BitConverter.UInt32BitsToSingle(0x7FC00000), 0 };
+                    yield return new object[] { BitConverter.UInt32BitsToSingle(0x7FC00000), BitConverter.UInt32BitsToSingle(0x7FC00001), 0 }; // implementation defined, not part of IEEE 754 totalOrder                    
+                }
```
Comment (tannergooding) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Numerics/TotalOrderIeee754ComparerTests.cs:116]: I'm not sure this is the desirable/best fix.

The overall preservation of a `NaN` sign/payload is implementation defined; however, in the face of a `NaN` with sign and/or payload, IEEE 754 `totalOrder` describes exact handling/behavior that is required to be followed.

The test was using `SingleTestData` for simplicity purposes; however in the face of `RiscV` it is likely better to fix the test to take `NFloat` and to enumerate bitwise correct values depending on the underlying bitness of the platform. That way encountered `NaN` with sign/payload (such as from the developer doing a bitcast where RiscV doesn't normalize) can be validated.

Comment (denis-paranichev) [reply src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Numerics/TotalOrderIeee754ComparerTests.cs:116]: Do you mean to enumerate bitwise correct values only for NaN values or for all test values?

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Numerics/TotalOrderIeee754ComparerTests.cs [hunk 2]
```diff
@@ -103,30 +103,32 @@ public static IEnumerable<object[]> NFloatTestData
...
-                    yield return new object[] { BitConverter.UInt32BitsToSingle(0x7FC00000), BitConverter.UInt32BitsToSingle(0x7FC00001), 0 }; // implementation defined, not part of IEEE 754 totalOrder                    
+                    yield return new object[] { BitConverter.UInt64BitsToDouble(0x7FF80000_00000000), 1.0d, 1 };
+                    yield return new object[] { BitConverter.UInt64BitsToDouble(0x7FF80000_00000000), NFloat.PositiveInfinity, 1 };
+                    yield return new object[] { BitConverter.UInt64BitsToDouble(0xFFF80000_00000000), NFloat.NegativeInfinity, -1 };
+                    yield return new object[] { BitConverter.UInt64BitsToDouble(0xFFF80000_00000000), -1.0d, -1 };
+                    yield return new object[] { BitConverter.UInt64BitsToDouble(0xFFF80000_00000000), BitConverter.UInt64BitsToDouble(0x7FF80000_00000000), -1 };
+                    yield return new object[] { BitConverter.UInt64BitsToDouble(0x7FF80000_00000000), BitConverter.UInt64BitsToDouble(0x7FF80000_00000001), -1 }; // implementation defined, not part of IEEE 754 totalOrder
```
Comment (tannergooding) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Numerics/TotalOrderIeee754ComparerTests.cs:115]: The comment here is incorrect, IEEE 754 `totalOrder` exactly defines the ordering including of NaNs.

A given implementation may not guarantee that `NaNs` are not normalized prior to `totalOrder`, but if a `NaN` of a particular shape is encountered in `totalOrder`, it does have an exact ordering in relation to a `NaN` of another shape. The same is true of some other APIs, such as `Abs` or `CopySign`.

Comment (denis-paranichev) [reply src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Numerics/TotalOrderIeee754ComparerTests.cs]: Done, thank you for clarification

**Review by sirntar — APPROVED** [review]

**Comment by tannergooding** [comment]
Looks like there are some CI failures related to this PR that need to be resolved still, specifically related to `NFloat`..

Failures look to be similar to `System.ArgumentException : Object of type 'System.Double' cannot be converted to type 'System.Runtime.InteropServices.NFloat'.` and so I expect you need to ensure the test data pre-casts the literal to `NFloat`

**Comment by denis-paranichev** [comment]
> Looks like there are some CI failures related to this PR that need to be resolved still, specifically related to `NFloat`..
> 
> Failures look to be similar to `System.ArgumentException : Object of type 'System.Double' cannot be converted to type 'System.Runtime.InteropServices.NFloat'.` and so I expect you need to ensure the test data pre-casts the literal to `NFloat`

Thank you for noticing, I have fixed this failure.

**Review by tannergooding — APPROVED** [review]

**Comment by denis-paranichev** [comment]
@tannergooding, thank you for merge it!

**Pull request merged**